### ISSUE TITLE
API Gateway Management API post to connection throws connection refused

### ISSUE THREAD

**Issue by m-radzikowski** [aws/aws-sdk-js-v3#999]
**Describe the bug**
I'm unable to send websocket message to API Gateway from Lambda using new v3 SDK.

**SDK version number**
"@aws-sdk/client-apigatewaymanagementapi": "1.0.0-alpha.22"

**Is the issue in the browser/Node.js?**
Node.js

**Details of the browser/Node.js version**
Lambda with Node v12.x

**To Reproduce (observed behavior)**
This code (TypeScript):
```typescript
import {
  ApiGatewayManagementApiClient,
  PostToConnectionCommand,
} from "@aws-sdk/client-apigatewaymanagementapi";

const apiGatewayManagementApi = new ApiGatewayManagementApiClient({
  apiVersion: "2018-11-29",
  endpoint: event.requestContext.domainName + "/" + event.requestContext.stage,
});

await apiGatewayManagementApi.send(
  new PostToConnectionCommand({
    // @ts-ignore
    Data: "hello from lambda!",
    ConnectionId: event.requestContext.connectionId!,
  })
);

```
throws an error:
```json
{
    "errorType": "Error",
    "errorMessage": "connect ECONNREFUSED 127.0.0.1:443",
    "code": "ECONNREFUSED",
    "errno": "ECONNREFUSED",
    "syscall": "connect",
    "address": "127.0.0.1",
    "port": 443,
    "$metadata": {
        "retries": 0,
        "totalRetryDelay": 0
    },
    "stack": [
        "Error: connect ECONNREFUSED 127.0.0.1:443",
        "    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1134:16)"
    ]
}
```

Moreover, I believe the type for the `Data` parameter in the command is wrong. It is Uint8Array, but when I provided data of this type, I get an error that "chunk expected string, got object". So I put string there and ignored type mismatch.

**Expected behavior**
With this code for the stable v2 aws-sdk (v2.639.0):
```typescript
import ApiGatewayManagementApi from "aws-sdk/clients/apigatewaymanagementapi";

const apiGatewayManagementApi = new ApiGatewayManagementApi({
  apiVersion: "2018-11-29",
  endpoint: event.requestContext.domainName + "/" + event.requestContext.stage,
});

await apiGatewayManagementApi
  .postToConnection({
    ConnectionId: event.requestContext.connectionId!,
    Data: "hello from lambda!",
  })
  .promise();

```
everything works.

**Comment by AllanFly120** [comment]
@m-radzikowski 
The failed request was sent to endpoint `127.0.0.1:443`. It means the the endpoint you supplied was not parsed correctly. Can you share an example of the endpoint you supplied? `event.requestContext.domainName + '/' + event.requestContext.stage`

**Comment by m-radzikowski** [comment]
The url is valid - I take it from the API Gateway event. And without any change in it it works with sdk v2. But I checked and logged full event and built url, just to be sure:

```json
{
    "_logLevel": "info",
    "msg": "Event",
    "obj": {
        "headers": {
            "Accept-Encoding": "gzip, deflate, br",
            "Accept-Language": "pl-PL,pl;q=0.9,en-US;q=0.8,en;q=0.7",
            "Cache-Control": "no-cache",
            "Host": "REMOVED.execute-api.eu-central-1.amazonaws.com",
            "Origin": "https://www.websocket.org",
            "Pragma": "no-cache",
            "Sec-WebSocket-Extensions": "permessage-deflate; client_max_window_bits",
            "Sec-WebSocket-Key": "REMOVED",
            "Sec-WebSocket-Version": "13",
            "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36",
            "X-Amzn-Trace-Id": "Root=1-5e70bc40-7c0fa1f010b406c84c091c90",
            "X-Forwarded-For": "REMOVED",
            "X-Forwarded-Port": "443",
            "X-Forwarded-Proto": "https"
        },
        "multiValueHeaders": {
            "Accept-Encoding": [
                "gzip, deflate, br"
            ],
            "Accept-Language": [
                "pl-PL,pl;q=0.9,en-US;q=0.8,en;q=0.7"
            ],
            "Cache-Control": [
                "no-cache"
            ],
            "Host": [
                "REMOVED.execute-api.eu-central-1.amazonaws.com"
            ],
            "Origin": [
                "https://www.websocket.org"
            ],
            "Pragma": [
                "no-cache"
            ],
            "Sec-WebSocket-Extensions": [
                "permessage-deflate; client_max_window_bits"
            ],
            "Sec-WebSocket-Key": [
                "REMOVED"
            ],
            "Sec-WebSocket-Version": [
                "13"
            ],
            "User-Agent": [
                "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36"
            ],
            "X-Amzn-Trace-Id": [
                "Root=1-5e70bc40-7c0fa1f010b406c84c091c90"
            ],
            "X-Forwarded-For": [
                "REMOVED"
            ],
            "X-Forwarded-Port": [
                "443"
            ],
            "X-Forwarded-Proto": [
                "https"
            ]
        },
        "queryStringParameters": {
            "encoding": "text"
        },
        "multiValueQueryStringParameters": {
            "encoding": [
                "text"
            ]
        },
        "requestContext": {
            "routeKey": "$connect",
            "messageId": null,
            "eventType": "CONNECT",
            "extendedRequestId": "JiJaHHxbliAFXrQ=",
            "requestTime": "17/Mar/2020:12:02:08 +0000",
            "messageDirection": "IN",
            "stage": "v1",
            "connectedAt": 1584446528689,
            "requestTimeEpoch": 1584446528690,
            "identity": {
                "cognitoIdentityPoolId": null,
                "cognitoIdentityId": null,
                "principalOrgId": null,
                "cognitoAuthenticationType": null,
                "userArn": null,
                "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.132 Safari/537.36",
                "accountId": null,
                "caller": null,
                "sourceIp": "89.64.88.183",
                "accessKey": null,
                "cognitoAuthenticationProvider": null,
                "user": null
            },
            "requestId": "JiJaHHxbliAFXrQ=",
            "domainName": "REMOVED.execute-api.eu-central-1.amazonaws.com",
            "connectionId": "JiJaHcx7FiACFOA=",
            "apiId": "REMOVED"
        },
        "isBase64Encoded": false
    },
    "awsRequestId": "729c1358-e983-4da3-ae1c-0934c5e5d3c1",
    "_tags": [
        "log",
        "info"
    ]
}
```

Built url:
```json
REMOVED.execute-api.eu-central-1.amazonaws.com/v1
```

Of course `REMOVED` is edited right now by me. So data provided is ok, it looks like the SDK ignores it and uses some default value, that is localhost.

**Comment by lukehedger** [comment]
@m-radzikowski Did you find the cause of this error? I am experiencing a similar error using the Secrets Manager client.

I initialise the client like so:
```js
import {
  GetSecretValueCommand,
  SecretsManagerClient,
} from "@aws-sdk/client-secrets-manager";

const secretsManager = new SecretsManagerClient({
  region: "eu-west-2",
});

const endpoint = await secretsManager.config.endpoint();

console.log(endpoint);

const getSecretValueCommand = new GetSecretValueCommand({
  SecretId: secretId,
});

const secretValue = await secretsManager.send(getSecretValueCommand); // throws
```

The endpoint config is returned as:
```js
{
  hostname: 'secretsmanager.eu-west-2.amazonaws.com',
  port: undefined,
  protocol: 'https:',
  path: '/',
  query: undefined
}
```

This is the error thrown:
```
Error: connect ECONNREFUSED 127.0.0.1:443
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1141:16) {
  errno: 'ECONNREFUSED',
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 443,
  '$metadata': { retries: 0, totalRetryDelay: 0 }
}
```

After running this under various conditions, I'm pretty convinced it only happens on the initial invocation of the Lambda the client is used in. On subsequent invocations the secret value is returned.

I was using version `1.0.0-beta.3` but upgraded to `1.0.0-gamma.2` to see if that made a difference - same result.

@AllanFly120 Do you have any idea of what could be causing this?

**Comment by m-radzikowski** [comment]
I did not find the cause. I think it's a bug in the sdk, and had to rollback to v2.

**Comment by lukehedger** [comment]
> I did not find the cause. I think it's a bug in the sdk, and had to rollback to v2.

Thanks @m-radzikowski

@AllanFly120 Are you able to confirm this is a known bug?

**Comment by gatsbyz** [comment]
Seeing the same for cloud-search-domain v3 (1.0.0-beta.5)

**Comment by zakkor** [comment]
I am experiencing the same issue using the S3 client

**Comment by trivikr** [comment]
@lukehedger @m-radzikowski Can you reproduce this issue in 1.0.0-gamma.3?
* `@aws-sdk/client-apigatewaymanagementapi@1.0.0-gamma.3`
* `@aws-sdk/client-secrets-manager@1.0.0-gamma.3`

**Comment by m-radzikowski** [comment]
@trivikr yes, the problem is still there on `1.0.0-gamma.3`.

Here I created a simple repo with bug reproduction: https://github.com/m-radzikowski/aws-sdk-js-v3-issue999

**Comment by mapoulos** [comment]
FWIW, I hit the same problem on `1.0.0-gamma.11`.

**Comment by 0c370t** [comment]
I also encountered this issue, was able to circumvent by passing in an object that implements the `Endpoint` interface. 

```typescript
        this.s3 = new S3({
            region: "[region]",
            endpoint: {
                protocol: "https",
                hostname: "[YOUR HOSTNAME]",
                path: ""
            },
        });
```

Full interface if you need more properties:
```typescript
export interface Endpoint {
  protocol: string;
  hostname: string;
  port?: number;
  path: string;
  query?: QueryParameterBag;
}
```

**Comment by trivikr** [comment]
Findings:
* If endpoint is of type `string`, it's parsed using urlParser in EndpointsConfig.ts https://github.com/aws/aws-sdk-js-v3/blob/bfbc78b3491e2b216fedbce8a2c09f0ff8fa40f2/packages/config-resolver/src/EndpointsConfig.ts#L38
* The urlParser is defined in `@aws-sdk/url-parser-node` which uses deprecated [url.parse](https://nodejs.org/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost): https://github.com/aws/aws-sdk-js-v3/blob/bfbc78b3491e2b216fedbce8a2c09f0ff8fa40f2/packages/url-parser-node/src/index.ts#L6

While parsing the string `REMOVED.execute-api.eu-central-1.amazonaws.com/v1`, the hostname is returned as null and is assigned `"localhost"` as shown below:

```console
$ node           
Welcome to Node.js v14.15.1.
Type ".help" for more information.
> process.version
'v14.15.1'
> const { parse } = require("url");
undefined
> parse("REMOVED.execute-api.eu-central-1.amazonaws.com/v1");
Url {
  protocol: null,
  slashes: null,
  auth: null,
  host: null,
  port: null,
  hostname: null,
  hash: null,
  search: null,
  query: null,
  pathname: 'REMOVED.execute-api.eu-central-1.amazonaws.com/v1',
  path: 'REMOVED.execute-api.eu-central-1.amazonaws.com/v1',
  href: 'REMOVED.execute-api.eu-central-1.amazonaws.com/v1'
}
```

**Comment by trivikr** [comment]
Code fix: Explicitly prefix `https://` when passing the endpoint.

```console
$ node           
Welcome to Node.js v14.15.1.
Type ".help" for more information.
> process.version
'v14.15.1'
> const { parse } = require("url");
undefined
> parse("https://REMOVED.execute-api.eu-central-1.amazonaws.com/v1");
Url {
  protocol: 'https:',
  slashes: true,
  auth: null,
  host: 'removed.execute-api.eu-central-1.amazonaws.com',
  port: null,
  hostname: 'removed.execute-api.eu-central-1.amazonaws.com',
  hash: null,
  search: null,
  query: null,
  pathname: '/v1',
  path: '/v1',
  href: 'https://removed.execute-api.eu-central-1.amazonaws.com/v1'
}
```

Code to update:
```diff
   const apiGatewayManagementApi = new ApiGatewayManagementApiClient({
     apiVersion: "2018-11-29",
     endpoint:
-      event.requestContext.domainName + "/" + event.requestContext.stage,
+      "https://" + event.requestContext.domainName + "/" + event.requestContext.stage,
   });
```

**Comment by m-radzikowski** [comment]
@trivikr thanks! This is a change from how the v2 SDK works, will it be changed in v3 to keep compatibility in this area or at least some info will be added to some "breaking changes" doc?

**Comment by trivikr** [comment]
@m-radzikowski we'll make a change to throw error when protocol is not passed by switching to WHATWG URL API.

```console
$ node           
Welcome to Node.js v14.15.1.
Type ".help" for more information.
> process.version
'v14.15.1'
> const { parse } = require("url");
undefined
> parse("https://REMOVED.execute-api.eu-central-1.amazonaws.com/v1");
Url {
  protocol: 'https:',
  slashes: true,
  auth: null,
  host: 'removed.execute-api.eu-central-1.amazonaws.com',
  port: null,
  hostname: 'removed.execute-api.eu-central-1.amazonaws.com',
  hash: null,
  search: null,
  query: null,
  pathname: '/v1',
  path: '/v1',
  href: 'https://removed.execute-api.eu-central-1.amazonaws.com/v1'
}
> new URL("REMOVED.execute-api.eu-central-1.amazonaws.com/v1");
Uncaught:
TypeError [ERR_INVALID_URL]: Invalid URL: REMOVED.execute-api.eu-central-1.amazonaws.com/v1
    at onParseError (internal/url.js:258:9)
    at new URL (internal/url.js:334:5)
    at REPL9:1:1
    at Script.runInThisContext (vm.js:132:18)
    at REPLServer.defaultEval (repl.js:484:29)
    at bound (domain.js:430:14)
    at REPLServer.runBound [as eval] (domain.js:443:12)
    at REPLServer.onLine (repl.js:817:10)
    at REPLServer.emit (events.js:327:22)
    at REPLServer.EventEmitter.emit (domain.js:486:12) {
  input: 'REMOVED.execute-api.eu-central-1.amazonaws.com/v1',
  code: 'ERR_INVALID_URL'
}
```

url.parse has been deprecated and it recommends using the WHATWG URL API
https://nodejs.org/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost

WHATWG URL has been available on Global Object since v10.0.0, the minimum Node.js version we support.
https://nodejs.org/api/url.html#url_the_whatwg_url_api
